pyomo/contrib/cp/transform/logical_to_disjunctive_walker.py

```diff
@@ -0,0 +1,235 @@
+    atmost_disj.constraint = Constraint(expr=sum_expr <= n)
+    more_disj.constraint = Constraint(expr=sum_expr >= n + 1)
+    visitor.disjunctions[len(visitor.disjunctions)] = [atmost_disj, more_disj]
+    return atmost_disj.indicator_var.get_associated_binary()
+
+_operator_dispatcher = {}
+_operator_dispatcher[LE.ImplicationExpression] = _dispatch_implication
+_operator_dispatcher[LE.EquivalenceExpression] = _dispatch_equivalence
+_operator_dispatcher[LE.NotExpression] = _dispatch_not
+_operator_dispatcher[LE.AndExpression] = _dispatch_and
+_operator_dispatcher[LE.OrExpression] = _dispatch_or
+_operator_dispatcher[LE.XorExpression] = _dispatch_xor
+_operator_dispatcher[LE.ExactlyExpression] = _dispatch_exactly
+_operator_dispatcher[LE.AtLeastExpression] = _dispatch_atleast
+_operator_dispatcher[LE.AtMostExpression] = _dispatch_atmost
+
+_before_child_dispatcher = {}
+_before_child_dispatcher[BV.ScalarBooleanVar] = _dispatch_boolean_var
+_before_child_dispatcher[BV._GeneralBooleanVarData] = _dispatch_boolean_var
+_before_child_dispatcher[AutoLinkedBooleanVar] = _dispatch_boolean_var
+_before_child_dispatcher[_ParamData] = _dispatch_param
+_before_child_dispatcher[ScalarParam] = _dispatch_param
+# for the moment, these are all just so we can get good error messages when we
+# don't handle them:
+_before_child_dispatcher[ScalarVar] = _dispatch_var
+_before_child_dispatcher[_GeneralVarData] = _dispatch_var
+_before_child_dispatcher[_GeneralExpressionData] = _dispatch_expression
+_before_child_dispatcher[ScalarExpression] = _dispatch_expression
+
+class LogicalToDisjunctiveVisitor(StreamBasedExpressionVisitor):
+    """Converts BooleanExpressions to Linear (MIP) representation
+
+    This converter eschews conjunctive normal form, and instead follows
+    the well-trodden MINLP path of factorable programming.
+
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.z_vars = VarList(domain=Binary)
+        self.z_vars.construct()
+        self.constraints = ConstraintList()
+        self.disjuncts = Disjunct(NonNegativeIntegers, concrete=True)
+        self.disjunctions = Disjunction(NonNegativeIntegers)
+        self.disjunctions.construct()
+        self.expansions = ComponentMap()
+        self.boolean_to_binary_map = ComponentMap()
+
+    def initializeWalker(self, expr):
+        walk, result = self.beforeChild(None, expr, 0)
+        if not walk:
+            return False, self.finalizeResult(result)
+        return True, expr
+
+    def beforeChild(self, node, child, child_idx):
+        if child.__class__ in EXPR.native_types:
+            return False, child
+
+        if isinstance(child, NumericExpression):
+            # Just pass it through, we'll figure it out later
+            return False, child
+        if isinstance(child, RelationalExpression):
+            # Eventually we'll handle these. Right now we set a MouseTrap
+            return _before_relational_expr(self, child)
```

Comment:
I think you should use `child.is_numeric_type()` and `child.is_expression_type(ExpressionType.RELATIONAL)` instead of `isinstance()` here (it is faster, and allows for duck-typing).